FIX: Handle channel dropout in eyelink files (e.g. switching from binocular to monocular mode etc)

scott-huberty opened the pull request:
This is an issue as PR. 

A file from one of our studies causes `read_raw_eyelink` to break. The file starts recording binocular data and switches to monocular data mid-way through the recording. This is something that EyeLink acquisition software allows.

Although I can't share the file, I’ve created a test that simulates the error. I'm still investigating a fix.

### Potential Solutions
- Add a Check for this scenario (binocular to monocular switch or vice versa) and raise a more informative error. The user can then regenerate their ASCII file, outputting monocular data only, which in theory should then be readable by MNE. I think this minimizes technical debt for us.
- Devise a workaround to handle this situation, such as adding a parameter to `read_raw_eyelink` that reads data for one eye only. This is probably (a lot?) more development work.
- Try to convince EyeLink (SR Research) to open-source a python reader for their format 😆 


CC @anomalosepia

scott-huberty committed:
TST: Make it break

scott-huberty committed:
Changelog [ci skip]

larsoner commented:
To me naively the easiest solution is actually to treat the data as binocular, but set one of the eyes to `nan` when it switches to monocular. Then if you want a "monocular instance" you just `raw.drop_channels(["eye_right"])` or whatever (assuming left is the one that was on the whole time).

scott-huberty commented:
> To me naively the easiest solution is actually to treat the data as binocular, but set one of the eyes to `nan` when it switches to monocular. Then if you want a "monocular instance" you just `raw.drop_channels(["eye_right"])` or whatever (assuming left is the one that was on the whole time).

Hmm yeah could do.. I'll give it a try and will see how much complexity it adds.

anomalosepia commented:
Hey Scott, hope you’re good.  any updates on this? I actually found a bunch of files where we have the opposite problem monocular to binocular.
Thanks!
a
From: Scott Huberty ***@***.***>
Date: Friday, September 13, 2024 at 10:43 AM
To: mne-tools/mne-python ***@***.***>
Cc: Adam Naples ***@***.***>, Mention ***@***.***>
Subject: Re: [mne-tools/mne-python] BUG: read raw eyelink file that switches from binocular to monocular mode (PR #12847)

To me naively the easiest solution is actually to treat the data as binocular, but set one of the eyes to nan when it switches to monocular. Then if you want a "monocular instance" you just raw.drop_channels(["eye_right"]) or whatever (assuming left is the one that was on the whole time).

Hmm yeah could do.. I'll give it a try and will see how much complexity it adds.

—
Reply to this email directly, view it on GitHub<https://github.com/mne-tools/mne-python/pull/12847#issuecomment-2349127215>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AB4VNG7OBPM3LMSO6MJX5KTZWL2ZDAVCNFSM6AAAAABOD6UXKKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGNBZGEZDOMRRGU>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

scott-huberty commented:
Hey @anomalosepia Thanks for the ping ; ) 

Someone else has reported the same issue.. I will try to get to it this weekend!

scott-huberty committed:
Merge remote-tracking branch 'upstream/main' into abcct_bug_fix

scott-huberty committed:
WIP, FIX: Make new test pass

Havent tested the rest of the suite yet.. Will likely have to make some more fixes due to the updated code

scott-huberty committed:
TST, FIX: make tests pass again

Actually we are now catching a case  that we didn't before, which is good! That is why I adjusted test_multi_block_misc_channels, to make sure we continue to warn when the eye being tracked changes

scott-huberty committed:
TST: add binocular velocity data to beef up codecov

scott-huberty committed:
STY: remove type hins in private function

- I just added them there for myself when I was debugging. But type hints dont play well with the _soft_import func
- also removed some cruft comments

scott-huberty committed:
FIX, TST: Better file handling practices suggested by Eric

Co-authored-by: Eric Larson <larson.eric.d@gmail.com>

scott-huberty committed:
FIX, STY: More informative messages on warnings and errors

- explicitly check that the sfreq didnt change during recording
- tell the user the implications of channel dropout

Co-authored-by: Eric Larson <larson.eric.d@gmail.com>

scott-huberty commented on mne/io/eyelink/tests/test_eyelink.py:
```diff
@@ -194,6 +194,115 @@ def test_find_overlaps():
+        second_block.append("\t".join(end_block))
+        lines += second_block
+        with out_file.open("w") as file:
+            file.writelines(lines)
```
Applied in [9d0ab84](https://github.com/mne-tools/mne-python/pull/12847/commits/9d0ab84fae85829d9fd3c2bf9fefb3ca4ba29c1a)

scott-huberty commented on mne/io/eyelink/_utils.py:
```diff
@@ -47,20 +47,27 @@ def _parse_eyelink_ascii(
+    for block in data_blocks:
+        del block["samples"]  # remove samples from block to save memory
+
+    # TODO: We still assume that sfreq and units are static across all blocks.
```
Ok addressed in [bed90d0](https://github.com/mne-tools/mne-python/pull/12847/commits/bed90d094f4e8ca45e2afe262d3dc588036aacd4)

I put this check with the other checks inside `_validate_data`

scott-huberty commented on mne/io/eyelink/_utils.py:
```diff
@@ -117,59 +124,85 @@ def _parse_recording_blocks(fname):
+        )
+    # Monocular tracking but switched between left/right eye
+    elif len(set(eyes)) > 1:
+        warn("The eye being tracked changed during the recording.")
```
addressed in [bed90d0](https://github.com/mne-tools/mne-python/pull/12847/commits/bed90d094f4e8ca45e2afe262d3dc588036aacd4)

scott-huberty committed:
TST: warning

scott-huberty committed:
Merge branch 'main' into abcct_bug_fix

scott-huberty commented:
Hey @larsoner can you re-run that failed azure job? It looks like it timed out.
